mixer/pkg/protobuf/yaml/dynamic/auth.go

```diff
@@ -0,0 +1,288 @@
+	// load public root cert.
+	publicCerts, err := ioutil.ReadFile(publicCACertsFile)
+	if err != nil {
+		log.Warnf("cannot load public root cert: %v", err)
+	} else {
+		caCertPool.AppendCertsFromPEM(publicCerts)
+	}
+
+	// load ca provided by file, such as istio CA certs.
+	if caFile != "" {
+		caCerts, err := ioutil.ReadFile(caFile)
+		if err != nil {
+			return nil, err
+		}
+		caCertPool.AppendCertsFromPEM(caCerts)
+	}
+	return caCertPool, nil
+}
+
+func getWhitelistSAN(cert []byte) []*url.URL {
+	var whitelistSAN []*url.URL
+	c, err := x509.ParseCertificate(cert)
+	if err != nil {
+		return whitelistSAN
+	}
+	for _, uri := range c.URIs {
+		whitelistSAN = append(whitelistSAN, uri)
+	}
+	return whitelistSAN
+}
+
+func buildMTLSDialOption(mtlsCfg *policypb.Mutual) ([]grpc.DialOption, error) {
+	// load peer cert/key.
+	pk := mtlsCfg.GetPrivateKey()
+	cc := mtlsCfg.GetClientCertificate()
+	peerCert, err := tls.LoadX509KeyPair(cc, pk)
+	if err != nil {
+		return nil, errors.Errorf("load peer cert/key error: %v", err)
+	}
+
+	// for simplicity, when using mtls, restrict out of process adapter cert subject to be mixer's spiffe identity
+	// TODO(bianpengyuan) make this configurable if needed
+	var whitelistSAN []*url.URL
+	if peerCert.Certificate != nil && len(peerCert.Certificate) > 0 {
+		whitelistSAN = getWhitelistSAN(peerCert.Certificate[0])
+	}
+
+	// Load certificates, include public certs and certs from config (such as istio CA certs).
+	caCertPool, err := loadCerts(mtlsCfg.GetCaCertificates())
+	if err != nil {
+		return nil, fmt.Errorf("failed to build TLS dial option: CA certificate cannot be loaded: %v", err)
+	}
+
+	customVerify := func(rawCerts [][]byte, _ [][]*x509.Certificate) error {
+		certs := make([]*x509.Certificate, len(rawCerts))
+		for i, asn1Data := range rawCerts {
+			cert, err := x509.ParseCertificate(asn1Data)
+			if err != nil {
+				return errors.New("failed to parse certificate from server: " + err.Error())
+			}
+			certs[i] = cert
+		}
+		opts := x509.VerifyOptions{
+			Roots:         caCertPool,
+			CurrentTime:   time.Now(),
+			Intermediates: x509.NewCertPool(),
+		}
+
+		for i, cert := range certs {
+			if i == 0 {
+				continue
+			}
+			opts.Intermediates.AddCert(cert)
+		}
+		if _, err = certs[0].Verify(opts); err != nil {
+			return nil
+		}
+		for _, uri := range certs[0].URIs {
+			for _, whitelisted := range whitelistSAN {
+				if reflect.DeepEqual(uri, whitelisted) {
+					return nil
+				}
+			}
+		}
+		return fmt.Errorf("failed to authenticate, cert SAN %v is not whitelisted", certs[0].URIs)
+	}
+
+	tc := credentials.NewTLS(&tls.Config{
+		Certificates: []tls.Certificate{peerCert},
+		RootCAs:      caCertPool,
+		// For mtls, skip default cert verification and use the custom one, which basically replicates the
+		// default verification logic. The only difference is that it checks subject alt name to be
+		// whitelisted spiffe URI instead of checking DNS/server name.
+		InsecureSkipVerify:    true,
```

Comment:
This should not be enabled. We still want chain verification from the root cert. Given that golang does not give us an option to separate host name verification from SNI, I suggest we punt on it, and force host verification. I glanced over godocs for TLS, but there seems to be no way to verify trust chain but skip host name field check.

Maybe @ozevren has some ideas, since he wrote similar code in Galley context...

Reply:
I think the chain certs are verified through this custom verification. The only difference between this custom logic and the default in go tls library is that it does not enforce dns or server name check (which will fail on istio mtls case since URI is not considered in SAN verification in go tls).

Reply:
If I remember correctly, galley service account is whitelisted in citadel to include common name in SAN, which will make it pass DNS name verification in mixer/pilot.